scripts/casefold_db.py

```diff
@@ -70,81 +161,110 @@ def update_local_associations(
-        db_update_args.append(
-            (
-                casefold_address,
-                assoc_tuples[0][2],
-                assoc_tuples[0][0],
-                assoc_tuples[0][1],
+        deltas[casefold_address] = Delta(
+            to_update=UpdateDelta(
+                mxid=assoc_tuples[0][0],
+                lookup_hash=assoc_tuples[0][1],
+                address=assoc_tuples[0][2],
             )
         )
 
         if len(assoc_tuples) > 1:
             # Iterate over all associations except for the first one, since we've already
             # processed it.
             for address, mxid, _ in assoc_tuples[1:]:
-                to_delete.append((address,))
-                mxids.append((mxid, address))
-
-    # iterate through the mxids and send email, let's only send one email per mxid
-    if send_email and not dry_run:
-        for mxid, address in mxids:
-            processed_mxids = []
-
-            if mxid in processed_mxids:
-                continue
-            else:
-                templateFile = sydent.get_branded_template(
-                    None,
-                    "migration_template.eml",
-                    ("email", "email.template"),
-                )
-
-                sendEmail(
-                    sydent,
-                    templateFile,
-                    address,
-                    {"mxid": "mxid", "subject_header_value": "MatrixID Update"},
+                deltas[casefold_address].to_delete.append(
+                    DeleteDelta(
+                        address=address,
+                        mxid=mxid,
+                    )
                 )
-                processed_mxids.append(mxid)
 
-    print(
-        f"{len(to_delete)} rows to delete, {len(db_update_args)} rows to update in local_threepid_associations"
-    )
+    if not test:
+        print(f"{len(deltas)} rows to update in local_threepid_associations")
 
-    if not dry_run:
-        if len(to_delete) > 0:
-            cur.executemany(
-                "DELETE FROM local_threepid_associations WHERE address = ?", to_delete
+    # Apply the deltas
+    for casefolded_address, delta in deltas.items():
+        if not test:
+            print(
+                f"Updating {casefolded_address} and deleting {len(delta.to_delete)} rows associated with it"
             )
 
-        if len(db_update_args) > 0:
-            cur.executemany(
-                "UPDATE local_threepid_associations SET address = ?, lookup_hash = ? WHERE address = ? AND mxid = ?",
-                db_update_args,
-            )
+        try:
+            # Delete each association, and send an email mentioning the affected MXID.
+            for to_delete in delta.to_delete:
+                cur = db.cursor()
+                if not dry_run:
+                    cur.execute(
+                        "DELETE FROM local_threepid_associations WHERE address = ?",
+                        (to_delete.address,),
+                    )
```

Comment:
I'd move this *after* the `sendEmailWithBackoff`, there's not much reason to do the query before then wait for a potentially long time before committing. (Doesn't matter with SQLite, but would matter a lot if using Postgres). It also means we don't need to worry about rolling back the transaction when we raise an exception.